redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+        self.from_url = from_url
+        self._skip_full_coverage_check = skip_full_coverage_check
+        self.reset_connections = False
+        self._moved_exception = None
+        self.connection_kwargs = kwargs
+        self.read_load_balancer = None
+        # Currently we do not support passing connection classes,
+        # we use ClusterConnection by default or SSLClusterConnection for SSL
+        self.initialize()
+
+    def get_node(self, node_name=None, host=None, port=None):
+        if node_name is None:
+            if not host or not port:
+                warnings.warn(
+                    "get_node requires one of the followings: "
+                    "1. node name "
+                    "2. host and port"
+                )
+                return None
+            if host == "localhost":
+                host = socket.gethostbyname(host)
+            node_name = get_node_name(host=host, port=port)
+        return self.nodes_cache.get(node_name)
+
+    def update_moved_exception(self, exception):
+        self._moved_exception = exception
+
+    def _update_moved_slots(self):
+        e = self._moved_exception
+        redirected_node = self.get_node(host=e.host, port=e.port)
+        if redirected_node:
+            if redirected_node.server_type is not PRIMARY:
+                # Update the node's server type
+                redirected_node.server_type = PRIMARY
+        else:
+            # This is a new node, we will add it to the nodes cache
+            redirected_node = ClusterNode(e.host, e.port, PRIMARY)
+            self.nodes_cache[redirected_node.name] = redirected_node
+        if redirected_node in self.slots_cache[e.slot_id]:
+            # The MOVED error resulted from a failover, and the new slot owner
+            # had previously been a replica.
+            old_primary = self.slots_cache[e.slot_id][0]
+            # Update the old primary to be a replica and add it to the end of
+            # the slot's node list
+            old_primary.server_type = REPLICA
+            self.slots_cache[e.slot_id].append(old_primary)
+            # Remove the old replica, which is now a primary, from the slot's
+            # node list
+            self.slots_cache[e.slot_id].remove(redirected_node)
+            # Override the old primary with the new one
+            self.slots_cache[e.slot_id][0] = redirected_node
+        else:
+            # The new slot owner is a new server, or a server from a different
+            # shard. We need to remove all current nodes from the slot's list
+            # (including replications) and add just the new node.
+            self.slots_cache[e.slot_id] = [redirected_node]
+        # Reset moved_exception
+        self._moved_exception = None
+
+    def get_node_from_slot(self, slot, read_from_replicas=False,
+                           server_type=None):
+        """
+        Gets a node that servers this hash slot
+        """
+        try:
```

Comment:
Can we shorten the try/catch or modify is so that the minimal code that triggers the KeyError does the raise?  Alternatively a series of has_key checks could do a raise SlotNotCoveredError.

Reply:
Changed:
```
    def get_node_from_slot(self, slot, read_from_replicas=False,
                           server_type=None):
        """
        Gets a node that servers this hash slot
        """
        if self._moved_exception:
            self._update_moved_slots()
        if read_from_replicas:
            # get the server index in a Round-Robin manner
            node_idx = self.get_read_load_balancer().get_server_index(
                slot, len(self.slots_cache[slot]))
        elif (
                server_type is None
                or server_type == PRIMARY
                or len(self.slots_cache[slot]) == 1
        ):
            # return a primary
            node_idx = 0
        else:
            # return a replica
            # randomly choose one of the replicas
            node_idx = random.randint(
                1, len(self.slots_cache[slot]) - 1)
        try:
            target_node = self.slots_cache[slot][node_idx]
        except KeyError:
            raise SlotNotCoveredError(
                'Slot "{0}" not covered by the cluster. '
                '"skip_full_coverage_check={1}"'.format(
                    slot, self._skip_full_coverage_check)
            )
        return target_node
```